File: weight_tracker.php
<?php
// weight_tracker.php — PHP версия

$dataFile = 'weight_data.json';

function loadEntries() {
    global $dataFile;
    if (file_exists($dataFile)) {
        $json = file_get_contents($dataFile);
        return json_decode($json, true) ?: [];
    }
    return [];
}

function saveEntries($entries) {
    global $dataFile;
    file_put_contents($dataFile, json_encode($entries, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE));
}

$entries = loadEntries();

function color($text, $code) {
    return "\033[{$code}m{$text}\033[0m";
}

function listAll($entries) {
    if (empty($entries)) {
        echo color("Нет записей.\n", '33');
        return;
    }
    printf(color("%-4s %-12s %-12s %-30s\n", '36'), "ID", "Дата", "Вес (кг)", "Заметки");
    echo str_repeat("-", 60) . "\n";
    foreach ($entries as $e) {
        printf("%-4d %-12s %-12.1f %-30s\n", $e['id'], $e['date'], $e['weight'], $e['notes'] ?? '');
    }
}

function getStats($entries) {
    if (empty($entries)) return null;
    $weights = array_column($entries, 'weight');
    usort($entries, function($a, $b) { return strcmp($a['date'], $b['date']); });
    $first = $entries[0]['weight'];
    $last = $entries[count($entries)-1]['weight'];
    $change = $last - $first;
    $sum = array_sum($weights);
    $avg = $sum / count($weights);
    $min = min($weights);
    $max = max($weights);
    $bmi = null;
    if ($last > 0) {
        $height = 1.75;
        $bmi = $last / ($height * $height);
    }
    return ['count' => count($weights), 'current' => $last, 'average' => $avg, 'min' => $min, 'max' => $max, 'change' => $change, 'bmi' => $bmi];
}

function printStats($entries) {
    $s = getStats($entries);
    if (!$s) {
        echo color("Нет данных для статистики.\n", '33');
        return;
    }
    echo color("📊 Статистика:\n", '36');
    echo "  Всего записей: {$s['count']}\n";
    echo "  Текущий вес: " . round($s['current'], 1) . " кг\n";
    echo "  Средний вес: " . round($s['average'], 1) . " кг\n";
    echo "  Минимальный вес: " . round($s['min'], 1) . " кг\n";
    echo "  Максимальный вес: " . round($s['max'], 1) . " кг\n";
    $changeColor = $s['change'] < 0 ? '32' : ($s['change'] > 0 ? '31' : '33');
    echo "  Изменение: " . color(($s['change'] > 0 ? '+' : '') . round($s['change'], 1) . " кг", $changeColor) . "\n";
    if ($s['bmi']) {
        $bmiColor = $s['bmi'] >= 18.5 && $s['bmi'] <= 24.9 ? '32' : ($s['bmi'] >= 25 && $s['bmi'] <= 29.9 ? '33' : '31');
        echo "  ИМТ: " . color(round($s['bmi'], 1), $bmiColor) . "\n";
    }
}

function exportCSV($entries, $filename = 'weight_data.csv') {
    if (empty($entries)) {
        echo color("Нет данных для экспорта.\n", '33');
        return;
    }
    $fp = fopen($filename, 'w');
    fputcsv($fp, ['ID', 'Дата', 'Вес (кг)', 'Заметки']);
    foreach ($entries as $e) {
        fputcsv($fp, [$e['id'], $e['date'], round($e['weight'], 1), $e['notes'] ?? '']);
    }
    fclose($fp);
    echo color("💾 Экспорт CSV: $filename\n", '32');
}

function exportJSON($entries, $filename = 'weight_data_export.json') {
    if (empty($entries)) {
        echo color("Нет данных для экспорта.\n", '33');
        return;
    }
    file_put_contents($filename, json_encode($entries, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE));
    echo color("💾 Экспорт JSON: $filename\n", '32');
}

function deleteEntry(&$entries, $id) {
    foreach ($entries as $i => $e) {
        if ($e['id'] == $id) {
            array_splice($entries, $i, 1);
            saveEntries($entries);
            return true;
        }
    }
    return false;
}

function editEntry(&$entries, $id, $field, $value) {
    foreach ($entries as &$e) {
        if ($e['id'] == $id) {
            switch ($field) {
                case 'date': $e['date'] = $value; break;
                case 'weight': $e['weight'] = (float)$value; break;
                case 'notes': $e['notes'] = $value; break;
                default: return false;
            }
            saveEntries($entries);
            return true;
        }
    }
    return false;
}

function filterByDate($entries, $start, $end) {
    $results = array_filter($entries, function($e) use ($start, $end) {
        return $e['date'] >= $start && $e['date'] <= $end;
    });
    if (empty($results)) {
        echo color("Нет записей за указанный период.\n", '33');
        return;
    }
    foreach ($results as $e) {
        echo "{$e['id']}: {$e['date']} | " . round($e['weight'], 1) . " кг | {$e['notes']}\n";
    }
}

function main() {
    global $entries;
    while (true) {
        echo "\n" . color("⚖️ Weight Tracker Pro (PHP)\n", '36');
        echo "1. Добавить запись\n";
        echo "2. Показать все записи\n";
        echo "3. Статистика\n";
        echo "4. Экспорт в CSV\n";
        echo "5. Экспорт в JSON\n";
        echo "6. Удалить запись\n";
        echo "7. Редактировать запись\n";
        echo "8. Фильтр по дате\n";
        echo "9. Выход\n";
        echo "Выберите действие: ";
        $choice = trim(fgets(STDIN));

        switch ($choice) {
            case '1':
                echo "Дата (ГГГГ-ММ-ДД, Enter для сегодня): ";
                $date = trim(fgets(STDIN));
                if (empty($date)) $date = date('Y-m-d');
                echo "Вес (кг): ";
                $weight = (float) trim(fgets(STDIN));
                echo "Заметки (опционально): ";
                $notes = trim(fgets(STDIN));
                $id = count($entries) + 1;
                $entries[] = ['id' => $id, 'date' => $date, 'weight' => $weight, 'notes' => $notes];
                saveEntries($entries);
                echo color("✅ Запись добавлена (ID: $id)\n", '32');
                break;
            case '2':
                listAll($entries);
                break;
            case '3':
                printStats($entries);
                break;
            case '4':
                echo "Имя CSV файла (по умолч. weight_data.csv): ";
                $filename = trim(fgets(STDIN));
                $filename = empty($filename) ? 'weight_data.csv' : $filename;
                exportCSV($entries, $filename);
                break;
            case '5':
                echo "Имя JSON файла (по умолч. weight_data_export.json): ";
                $filename = trim(fgets(STDIN));
                $filename = empty($filename) ? 'weight_data_export.json' : $filename;
                exportJSON($entries, $filename);
                break;
            case '6':
                listAll($entries);
                echo "Введите ID для удаления: ";
                $id = (int) trim(fgets(STDIN));
                if (deleteEntry($entries, $id)) {
                    echo color("✅ Запись удалена.\n", '32');
                } else {
                    echo color("❌ Запись не найдена.\n", '31');
                }
                break;
            case '7':
                listAll($entries);
                echo "Введите ID для редактирования: ";
                $id = (int) trim(fgets(STDIN));
                echo "Какое поле редактировать (date, weight, notes): ";
                $field = trim(fgets(STDIN));
                echo "Новое значение: ";
                $value = trim(fgets(STDIN));
                if (editEntry($entries, $id, $field, $value)) {
                    echo color("✅ Запись обновлена.\n", '32');
                } else {
                    echo color("❌ Не удалось обновить.\n", '31');
                }
                break;
            case '8':
                echo "Начальная дата (ГГГГ-ММ-ДД): ";
                $start = trim(fgets(STDIN));
                echo "Конечная дата (ГГГГ-ММ-ДД): ";
                $end = trim(fgets(STDIN));
                filterByDate($entries, $start, $end);
                break;
            case '9':
                echo "До свидания!\n";
                exit(0);
            default:
                echo color("Неверный выбор.\n", '31');
        }
    }
}

main();
?>

File: weight_tracker.java
// weight_tracker.java — Java версия

import java.io.*;
import java.nio.file.*;
import java.time.*;
import java.util.*;

class WeightEntry {
    int id;
    String date;
    double weight;
    String notes;

    WeightEntry(int id, String date, double weight, String notes) {
        this.id = id;
        this.date = date;
        this.weight = weight;
        this.notes = notes;
    }

    String toJson() {
        return String.format("{\"id\":%d,\"date\":\"%s\",\"weight\":%.1f,\"notes\":\"%s\"}",
                id, date, weight, notes != null ? notes : "");
    }
}

public class weight_tracker {
    private static List<WeightEntry> entries = new ArrayList<>();
    private static final String DATA_FILE = "weight_data.json";
    private static Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) {
        load();
        while (true) {
            System.out.println("\n\u001B[36m⚖️ Weight Tracker Pro (Java)\u001B[0m");
            System.out.println("1. Добавить запись");
            System.out.println("2. Показать все записи");
            System.out.println("3. Статистика");
            System.out.println("4. Экспорт в CSV");
            System.out.println("5. Экспорт в JSON");
            System.out.println("6. Удалить запись");
            System.out.println("7. Редактировать запись");
            System.out.println("8. Фильтр по дате");
            System.out.println("9. Выход");
            System.out.print("Выберите действие: ");
            String choice = scanner.nextLine();
            switch (choice) {
                case "1": addEntry(); break;
                case "2": listAll(); break;
                case "3": printStats(); break;
                case "4": exportCSV(); break;
                case "5": exportJSON(); break;
                case "6": deleteEntry(); break;
                case "7": editEntry(); break;
                case "8": filterByDate(); break;
                case "9": System.out.println("До свидания!"); return;
                default: System.out.println("\u001B[31mНеверный выбор.\u001B[0m");
            }
        }
    }

    private static void load() {
        try {
            String content = new String(Files.readAllBytes(Paths.get(DATA_FILE)));
            // Упрощённо, для демонстрации используем пустой список
            entries = new ArrayList<>();
        } catch (IOException e) {
            entries = new ArrayList<>();
        }
    }

    private static void save() {
        try {
            StringBuilder sb = new StringBuilder("[");
            for (int i = 0; i < entries.size(); i++) {
                sb.append(entries.get(i).toJson());
                if (i < entries.size() - 1) sb.append(",");
            }
            sb.append("]");
            Files.write(Paths.get(DATA_FILE), sb.toString().getBytes());
        } catch (IOException e) {
            System.out.println("Ошибка сохранения.");
        }
    }

    private static void addEntry() {
        System.out.print("Дата (ГГГГ-ММ-ДД, Enter для сегодня): ");
        String date = scanner.nextLine();
        if (date.isEmpty()) {
            date = LocalDate.now().toString();
        }
        System.out.print("Вес (кг): ");
        double weight = Double.parseDouble(scanner.nextLine());
        System.out.print("Заметки (опционально): ");
        String notes = scanner.nextLine();
        int id = entries.size() + 1;
        entries.add(new WeightEntry(id, date, weight, notes));
        save();
        System.out.println("\u001B[32m✅ Запись добавлена (ID: " + id + ")\u001B[0m");
    }

    private static void listAll() {
        if (entries.isEmpty()) {
            System.out.println("\u001B[33mНет записей.\u001B[0m");
            return;
        }
        System.out.printf("\u001B[36m%-4s %-12s %-12s %-30s\u001B[0m\n", "ID", "Дата", "Вес (кг)", "Заметки");
        System.out.println("-".repeat(60));
        for (WeightEntry e : entries) {
            System.out.printf("%-4d %-12s %-12.1f %-30s\n", e.id, e.date, e.weight, e.notes);
        }
    }

    private static void printStats() {
        if (entries.isEmpty()) {
            System.out.println("\u001B[33mНет данных для статистики.\u001B[0m");
            return;
        }
        double[] weights = entries.stream().mapToDouble(e -> e.weight).toArray();
        entries.sort(Comparator.comparing(e -> e.date));
        double first = entries.get(0).weight;
        double last = entries.get(entries.size()-1).weight;
        double change = last - first;
        double sum = 0;
        double min = weights[0];
        double max = weights[0];
        for (double w : weights) {
            sum += w;
            if (w < min) min = w;
            if (w > max) max = w;
        }
        double avg = sum / weights.length;
        Double bmi = null;
        if (last > 0) {
            double height = 1.75;
            bmi = last / (height * height);
        }
        System.out.println("\u001B[36m📊 Статистика:\u001B[0m");
        System.out.printf("  Всего записей: %d\n", weights.length);
        System.out.printf("  Текущий вес: %.1f кг\n", last);
        System.out.printf("  Средний вес: %.1f кг\n", avg);
        System.out.printf("  Минимальный вес: %.1f кг\n", min);
        System.out.printf("  Максимальный вес: %.1f кг\n", max);
        String changeColor = change < 0 ? "\u001B[32m" : change > 0 ? "\u001B[31m" : "\u001B[33m";
        System.out.printf("  Изменение: %s%+.1f кг\u001B[0m\n", changeColor, change);
        if (bmi != null) {
            String bmiColor = bmi >= 18.5 && bmi <= 24.9 ? "\u001B[32m" : bmi >= 25 && bmi <= 29.9 ? "\u001B[33m" : "\u001B[31m";
            System.out.printf("  ИМТ: %s%.1f\u001B[0m\n", bmiColor, bmi);
        }
    }

    private static void exportCSV() {
        if (entries.isEmpty()) {
            System.out.println("\u001B[33mНет данных для экспорта.\u001B[0m");
            return;
        }
        try (FileWriter fw = new FileWriter("weight_data.csv")) {
            fw.write("ID,Дата,Вес (кг),Заметки\n");
            for (WeightEntry e : entries) {
                fw.write(e.id + "," + e.date + "," + String.format("%.1f", e.weight) + "," + e.notes + "\n");
            }
            System.out.println("\u001B[32m💾 Экспорт CSV: weight_data.csv\u001B[0m");
        } catch (IOException e) {
            System.out.println("Ошибка экспорта.");
        }
    }

    private static void exportJSON() {
        if (entries.isEmpty()) {
            System.out.println("\u001B[33mНет данных для экспорта.\u001B[0m");
            return;
        }
        try {
            StringBuilder sb = new StringBuilder("[");
            for (int i = 0; i < entries.size(); i++) {
                sb.append(entries.get(i).toJson());
                if (i < entries.size() - 1) sb.append(",");
            }
            sb.append("]");
            Files.write(Paths.get("weight_data_export.json"), sb.toString().getBytes());
            System.out.println("\u001B[32m💾 Экспорт JSON: weight_data_export.json\u001B[0m");
        } catch (IOException e) {
            System.out.println("Ошибка экспорта.");
        }
    }

    private static void deleteEntry() {
        listAll();
        System.out.print("Введите ID для удаления: ");
        int id = Integer.parseInt(scanner.nextLine());
        Iterator<WeightEntry> it = entries.iterator();
        while (it.hasNext()) {
            if (it.next().id == id) {
                it.remove();
                save();
                System.out.println("\u001B[32m✅ Запись удалена.\u001B[0m");
                return;
            }
        }
        System.out.println("\u001B[31m❌ Запись не найдена.\u001B[0m");
    }

    private static void editEntry() {
        listAll();
        System.out.print("Введите ID для редактирования: ");
        int id = Integer.parseInt(scanner.nextLine());
        WeightEntry target = null;
        for (WeightEntry e : entries) {
            if (e.id == id) {
                target = e;
                break;
            }
        }
        if (target == null) {
            System.out.println("\u001B[31m❌ Запись не найдена.\u001B[0m");
            return;
        }
        System.out.print("Какое поле редактировать (date, weight, notes): ");
        String field = scanner.nextLine().toLowerCase();
        System.out.print("Новое значение: ");
        String value = scanner.nextLine();
        boolean ok = true;
        switch (field) {
            case "date": target.date = value; break;
            case "weight": target.weight = Double.parseDouble(value); break;
            case "notes": target.notes = value; break;
            default: ok = false;
        }
        if (ok) {
            save();
            System.out.println("\u001B[32m✅ Запись обновлена.\u001B[0m");
        } else {
            System.out.println("\u001B[31m❌ Не удалось обновить.\u001B[0m");
        }
    }

    private static void filterByDate() {
        System.out.print("Начальная дата (ГГГГ-ММ-ДД): ");
        String start = scanner.nextLine();
        System.out.print("Конечная дата (ГГГГ-ММ-ДД): ");
        String end = scanner.nextLine();
        List<WeightEntry> results = new ArrayList<>();
        for (WeightEntry e : entries) {
            if (e.date.compareTo(start) >= 0 && e.date.compareTo(end) <= 0) {
                results.add(e);
            }
        }
        if (results.isEmpty()) {
            System.out.println("\u001B[33mНет записей за указанный период.\u001B[0m");
            return;
        }
        for (WeightEntry e : results) {
            System.out.printf("%d: %s | %.1f кг | %s\n", e.id, e.date, e.weight, e.notes);
        }
    }
}
